### DISCUSSION TITLE
Copernicus API on Branch 10

### DISCUSSION THREAD

**Discussion by SimonFisher92** [SimonFisher92/Scottish_Snow_Satellite_Data#23] [General]
@ipoole @murraycutforth 

I made good progress migrating to copernicus today on this branch

https://github.com/SimonFisher92/Scottish_Snow/tree/10-new_copernicus_api_trial

There are still some things im not happy with (image quality is one), but this is a nice downloader and avoids the LTA. Can fetch a patches data in about 2 min, has rudimentary saving too

**Comment by SimonFisher92** [comment]
Its been awesome to see this working! I watched my favourite patch appear in the folder in real-time

![image](https://github.com/SimonFisher92/Scottish_Snow/assets/98879661/68c43c7e-96be-4b72-96a5-93ceff3d7902)

**Comment by SimonFisher92** [comment]
Unfortunately, the resolution when downloaded (right) is not as good as the sentinel playground (left)

![image](https://github.com/SimonFisher92/Scottish_Snow/assets/98879661/45ad5686-9a2f-48fe-828e-6c8dd36c14bc)

**Reply by ipoole** [reply]
@SimonFisher92 , has the download command changed for this branch.  I tried same command as before, after ~ 1 minute I get a  "Max retries ... Connection timeout".    My bad, not understanding what you're doing here.  Could we have a quick call, this evening (Thursday) perhaps, or Friday to get me back on track?

**Reply by murraycutforth** [reply]
Nice one @SimonFisher92 ! From your example up there, I wonder if the resolution really is worse, or is it just that the sentinel playground applies some smoothing / interpolation?

**Comment by murraycutforth** [comment]
Just wanted to document my thinking on the patch size which we use for these downloads.

Apparently by the definition of the metre (https://earthscience.stackexchange.com/questions/6843/how-much-distance-does-0-001-circ-in-latitude-or-longitude-represent) there are 10,000,000 / 90 = 111,111.111 metres per degree of latitude, or 9e-6 degrees per metre. I've experimented with bounding boxes edges from 1km to 3km, and I think 1.5km is a happy medium:

![image](https://github.com/SimonFisher92/Scottish_Snow/assets/11088372/b42c6ebe-3573-4115-8c18-c4520a224e9f)

Smaller boxes make it hard to see what's happening, and larger boxes start overlapping - e.g. coire domhain is only 1km away from other big patches like coire an t-sneachda and coire an lochain.

This is implemented in src/produce_geojsons/produce_geojsons.py

**Reply by ipoole** [reply]
Of course, converting degrees of *longitude* to meters more tricky.   Well, just dependent on latitude.

**Reply by murraycutforth** [reply]
Indeed! My solution to this is to define the offset in terms of degrees of latitude (since that is constant), and then I have a function in the code which computes the longitude offset such that the bounding box is approximately square. It basically just comes down to dividing the longitude length in degrees by cos(latitude) I think.

**Comment by murraycutforth** [comment]
In terms of downloading all bands, there is a handy example here which I'm going to try to follow: https://sentinelhub-py.readthedocs.io/en/latest/examples/process_request.html#Example-3:-All-Sentinel-2's-raw-band-values

**Comment by murraycutforth** [comment]
Just a heads up, I've just run out of credits when trying to download all patches all at once.

Server response: "{"code": 403, "description": "Insufficient processing units or requests available in your account. Upgrade account or acquire additional credits."}"

I only managed to download 2 years of data before hitting the monthly requests limit!

It would be good to ask if they will give us some extra credits for this project, any tips on how you went about that with amazon @SimonFisher92 ?
